src/test/kotlin/app/batch/S3ReaderTest.kt

```diff
@@ -0,0 +1,152 @@
+
+    @MockBean
+    private lateinit var httpClientProvider: HttpClientProvider
+
+    @Before
+    fun setUp() {
+
+        s3Reader.reset()
+        Mockito.reset(s3Client)
+    }
+
+    @Test
+    fun should_Read_Data_And_Metadata_Files_In_A_Given_Prefix() {
+        listObjectsV2Result = ListObjectsV2Result()
+        listObjectsV2Result.prefix = S3_PREFIX_FOLDER
+
+        s3ObjectSummary1 = S3ObjectSummary()
+        s3ObjectSummary1.bucketName = BUCKET_NAME1
+        s3ObjectSummary1.key = VALID_DATA_KEY
+
+        s3ObjectSummary2 = S3ObjectSummary()
+        s3ObjectSummary2.bucketName = BUCKET_NAME1
+        s3ObjectSummary2.key = VALID_METADATA_KEY
+
+        listObjectsV2Result.objectSummaries.add(s3ObjectSummary1)
+        listObjectsV2Result.objectSummaries.add(s3ObjectSummary2)
+
+        s3Object1 = S3Object()
+        s3Object1.bucketName = BUCKET_NAME1
+        s3Object1.key = VALID_DATA_KEY
+        s3Object1.objectContent = S3ObjectInputStream(ByteArrayInputStream(OBJECT_CONTENT1.toByteArray()), HttpGet())
+
+        s3Object2 = S3Object()
+        s3Object2.bucketName = BUCKET_NAME1
+        s3Object2.key = VALID_METADATA_KEY
+        s3Object2.objectContent = S3ObjectInputStream(ByteArrayInputStream(OBJECT_CONTENT2.toByteArray()), HttpGet())
+
+        given(s3Client.listObjectsV2(BUCKET_NAME1, S3_PREFIX_FOLDER)).willReturn(listObjectsV2Result)
+        given(s3Client.getObject(BUCKET_NAME1, VALID_DATA_KEY)).willReturn(s3Object1)
+        given(s3Client.getObject(BUCKET_NAME1, VALID_METADATA_KEY)).willReturn(s3Object2)
+
+        val encryptedStream1 = s3Reader.read()
+        val dataStream = encryptedStream1?.dataInputStream
+        val metadataStream = encryptedStream1?.metadataInputStream
+
+        assertObjectContent(OBJECT_CONTENT1, dataStream)
+        assertObjectContent(OBJECT_CONTENT2, metadataStream)
+    }
+
+    @Test(expected = RuntimeException::class)
+    fun should_Throw_Exception_When_Metadata_File_Missing_In_A_Given_Prefix() {
+
+        listObjectsV2Result = ListObjectsV2Result()
+        listObjectsV2Result.prefix = S3_PREFIX_FOLDER
+
+        s3ObjectSummary1 = S3ObjectSummary()
+        s3ObjectSummary1.bucketName = BUCKET_NAME1
+        s3ObjectSummary1.key = VALID_DATA_KEY
+
+
+        listObjectsV2Result.objectSummaries.add(s3ObjectSummary1)
+
+        s3Object1 = S3Object()
+        s3Object1.bucketName = BUCKET_NAME1
+        s3Object1.key = VALID_DATA_KEY
+        s3Object1.objectContent = S3ObjectInputStream(ByteArrayInputStream(OBJECT_CONTENT1.toByteArray()), HttpGet())
+
+
+        given(s3Client.listObjectsV2(BUCKET_NAME1, S3_PREFIX_FOLDER)).willReturn(listObjectsV2Result)
+        given(s3Client.getObject(BUCKET_NAME1, VALID_DATA_KEY)).willReturn(s3Object1)
+
+        given(s3Client.listObjectsV2(BUCKET_NAME1, S3_PREFIX_FOLDER)).willReturn(listObjectsV2Result)
+        given(s3Client.getObject(BUCKET_NAME1, VALID_DATA_KEY)).willReturn(s3Object1)
+
+        s3Reader.read()
```

Comment:
Can we please assert the exception message, to be sure we are not leaking data, and what will be in our logs is useful?

Reply:
I have delegated all the pairing logic to keypairgenerator and all logs are asserted in its corresponding test keypairgeneratortest